state-chain/pallets/cf-broadcast/src/lib.rs

```diff
@@ -405,7 +408,25 @@ pub mod pallet {
 			// We treat a time out here as a Broadcast Failure. This is handled the same way - the
 			// current broadcaster is reported as Failed to broadcast, and a new broadcaster is
 			// nominated. If there are no more broadcaster available, then the broadcast is aborted.
-			let mut expiries = Timeouts::<T, I>::take(block_number);
+
+			let current_chain_block = T::ChainTracking::get_block_height();
+
+			// take all broadcasts which have timed out. Since iterators break if we update
+			// the map during iteration, we do this in two steps: first collect all expired_keys
+			// and separately the expired values, after that iterate again to delete all collected
+			// keys from storage.
+			let mut expired_keys = Vec::new();
+			let mut expiries = BTreeSet::new();
+			for (ix, val) in Timeouts::<T, I>::iter() {
```

Comment:
I don't like having to iterate over a map of all timeouts on every block... maybe it would be easier to just store this as a single `Vec<(BlockNumber, BTreeMap<..>)>` value?

Reply:
by BlockNumber you mean ChainBlockNumber?

Reply:
if yes - then yeah, looking now we only append to timeouts so writing is still cheap

Reply:
Yeah - To allow appending it might have to be a `Vec<(ChainBlockNumber, BroadcastId, ValidatorId)>`.

Reply:
Ok, it seems that @kylezs agrees with you on this, so I'm gonna try your suggestion.

Reply:
The reason to not use a BTreeSet is that you can use StorageAppend to just `::append(..)` the value efficiently to the end of the Vec (ie. just write a single element). Otherwise you need to read, decode, update, encode, write the whole Vec.

Reply:
Ok i see what you mean. tbh. I don't see a great reason to use a BTreeSet in storage at all. We need to iterate all the values anyway most of the time, and when we append we almost always append a single value (the only exception to this is during safe mode).